haystack/utils/augment_squad.py

```diff
@@ -0,0 +1,178 @@
+"""
+Script to perform data augmentation on a SQuAD like dataset to increase training data. It follows the approach oultined in the TinyBERT paper.
+Usage:
+    python augment_squad.py --squad_path <squad_path> --output_path <output_patn> \
+        --multiplication_factor <multiplication_factor> --word_possibilities <word_possibilities> \
+        --replace_probability <replace_probability>
+Arguments:
+    squad_path: Path to the input dataset. Must have the same structure as the official squad json.
+    output_path: Path to the output dataset.
+    multiplication_factor: Number of times to augment the dataset.
+    word_possibilities: Number of possible words to replace a word with.
+    replace_probability: Probability of replacing a word with a different word.
+"""
+
+
+import torch
+from transformers import BertForMaskedLM, BertTokenizer
+from copy import copy, deepcopy
+from pathlib import Path
+import requests
+from zipfile import ZipFile
+import numpy as np
+import random
+import argparse
+import json
+import logging
+from tqdm import tqdm
+
+logger = logging.getLogger(__name__)
+
+def load_glove(glove_path: Path = Path("glove.txt"), vocab_size: int = 100_000):
+    if not glove_path.exists():
+        zip_path = glove_path.parent / (glove_path.name + ".zip")
+        request = requests.get("https://nlp.stanford.edu/data/glove.42B.300d.zip", allow_redirects=True)
```

Comment:
Let's compare performance of https://nlp.stanford.edu/data/glove.840B.300d.zip and https://nlp.stanford.edu/data/glove.6B.zip 
It would be also interesting to see whether fasttext performs better than glove when used for data augmentation. We need a way to use data augmentation for non-English datasets as well and fasttext could help with that.
```
import fasttext
ft = fasttext.load_model('cc.en.300.bin') 
#german model: https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.de.300.bin.gz
ft.get_nearest_neighbors('hello')
```

Reply:
I agree that this would be interesting to see. I think with a model trained using whole word masking we could perhaps even try not using this additional step at all, however, it would probably be quite costly to test this.

Distilling SQuAD the way they did for the original results would be equivalent timewise to about 100 epochs on an unaugmented dataset. 100 of those epochs would take about 100*45min=75h. In addition to that, data augmentation takes about 45h because you need to do a forward pass for each word.

Although I am trying to improve data augmentation speed and we could use a smaller model as student, trying out different tokenizer is probably not worth it as only about 3,4% of all words are replaced using glove (tested with [this dataset](https://github.com/deepset-ai/haystack/blob/master/test/samples/squad/tiny.json)).